### DISCUSSION TITLE
why the mouse looks smoother than the screen

### DISCUSSION THREAD

**Discussion by 407219125** [FreeRDP/FreeRDP#8951] [Q&A]
As the title says，why the mouse looks smoother than the screen

**Comment by pnowack** [comment]
I assume your question is related to mouse pointer bitmaps included in the stream of a monitor.

Well, it's quite easy: There are two ways to handle the mouse pointer bitmap in a remote desktop session via RDP.

1. The server includes the mouse pointer inside the visible stream.
2. The server tells the client side the mouse pointer bitmap, that the client should use.

Way 1 has the disadvantage, that the mouse pointer would visibly lag behind the its actual position, where the user moved it, leading to a degraded experience.
This is because screen updates are large may involve large areas, especially when moving a window.

With way 2, the mouse pointer behaves like moving the mouse pointer in local applications on the client side. The absolute cursor position and when clicks happen are still forwarded to the server. But the overall experience feels a lot smoother, because all what the client does is set the mouse pointer bitmap to what the server side told the client side and that's it.

This is visible, when moving windows for example. The mouse pointer bitmap is small, may even be already cached (so the server then just tells the client in which cache slot the cursor bitmap to use is (very tiny update)).
The actual screen update is large, needs to be encoded, transmitted too (may be slowed down by the available bandwidth), decoded on the client side and displayed.
And that takes much more time.

Most RDP servers use for the better experience way 2, that's why moving the mouse pointer (e.g. when moving windows with the mouse), looks smoother than the actual screen can be updated.

A similar experience can be observed with games that use a software cursor vs games that use hardware cursor. If the PC cannot keep up with rendering and is very slow, the former one will impact the cursor behaviour.
The latter does not, the frame rendering is still slow, but won't impact the cursor at all, although the system cannot keep up with the rendering (leading still to a better experience than with a software cursor).
